sdk/core/System.ClientModel/src/Internal/StreamExtensions.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Internal;
+
+internal static class StreamExtensions
+{
+    public static async Task WriteAsync(this Stream stream, ReadOnlyMemory<byte> buffer, CancellationToken cancellation = default)
+    {
+        ClientUtilities.AssertNotNull(stream, nameof(stream));
+#if NETCOREAPP
+        await stream.WriteAsync(buffer, cancellation).ConfigureAwait(false);
+#else
+
+            if (buffer.Length == 0)
+                return;
+            byte[]? array = null;
+            try
+            {
+                if (MemoryMarshal.TryGetArray(buffer, out ArraySegment<byte> arraySegment))
+                {
+                    Debug.Assert(arraySegment.Array != null);
+                    await stream.WriteAsync(arraySegment.Array, arraySegment.Offset, arraySegment.Count, cancellation).ConfigureAwait(false);
+                }
+                else
+                {
+                    array = ArrayPool<byte>.Shared.Rent(buffer.Length);
+
+                    if (!buffer.TryCopyTo(array))
+                        throw new Exception("could not rent large enough buffer.");
+                    await stream.WriteAsync(array, 0, buffer.Length, cancellation).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                if (array != null)
+                    ArrayPool<byte>.Shared.Return(array);
```

Comment:
Is there a chance these buffers could contain sensitive data? If so, we might want to set the `clearArray` param to `true` here.

Reply:
This is code I [copied in from Azure.Core](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Internal/AzureBaseBuffersExtensions.cs#L14), so if we wanted to change it here, I would want us to change it in both places.  Are we ok making a functional change to Azure.Core-based clients that use  this functionality as well?  Might there be any risk of breaking changes in doing that?

Reply:
There should be no breaking change, this only ensures that arrays returned to the pool don't still contain data that can be read by a future renter. There is a perf cost, however. I'll do a bit of research here to gauge whether this is a risk worth worrying about.

Reply:
Very cool!  Let me know what you learn and we can always file an issue and loop back on this if it looks like something we'd rather come back to and spend more time getting right later!

Reply:
Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41046